django/db/models/sql/query.py

```diff
@@ -1217,17 +1217,15 @@ def build_filter(self, filter_expr, branch_negated=False, current_negated=False,
             raise FieldError("Cannot parse keyword query %r" % arg)
         lookups, parts, reffed_expression = self.solve_lookup_type(arg)
 
-        if not getattr(reffed_expression, 'filterable', True):
-            raise NotSupportedError(
-                reffed_expression.__class__.__name__ + ' is disallowed in '
-                'the filter clause.'
-            )
-
         if not allow_joins and len(parts) > 1:
             raise FieldError("Joined field references are not permitted in this query")
 
         pre_joins = self.alias_refcount.copy()
         value = self.resolve_lookup_value(value, can_reuse, allow_joins, simple_col)
+
+        if not getattr(reffed_expression, 'filterable', True) or not getattr(value, 'filterable', True):
+            raise NotSupportedError('Window is disallowed in the filter clause.')
+
```

Comment:
We can treat expressions with "`Window`" as a special case (to keep current messages) but we should also be prepared for custom expression that are not filterable, e.g.
```diff
diff --git a/django/db/models/sql/query.py b/django/db/models/sql/query.py
index 4ef25023fc..112edd6138 100644
--- a/django/db/models/sql/query.py
+++ b/django/db/models/sql/query.py
@@ -1223,8 +1223,12 @@ class Query(BaseExpression):
         pre_joins = self.alias_refcount.copy()
         value = self.resolve_lookup_value(value, can_reuse, allow_joins, simple_col)
 
-        if not getattr(reffed_expression, 'filterable', True) or not getattr(value, 'filterable', True):
-            raise NotSupportedError('Window is disallowed in the filter clause.')
+        for expr in (reffed_expression, value):
+            if not getattr(expr, 'filterable', True):
+                raise NotSupportedError(
+                    '%s is disallowed in the filter clause.'
+                    % ('Window' if expr.contains_over_clause else repr(expr))
+                )
 
         used_joins = {k for k, v in self.alias_refcount.items() if v > pre_joins.get(k, 0)}
```

> Also do you think filterable should be deprecated if/when #28333 is implemented?

`filterable` is a part of public API, so custom expression can you use. IMO we will not remove it even after [28333](https://code.djangoproject.com/ticket/28333) implementation.